File: agents/models/events.py
from pydantic import BaseModel, Field
from enum import Enum
from time import time
from uuid import uuid4


class AgentStatus(str, Enum):
    IDLE = "idle"
    WORKING = "working"
    DONE = "done"
    ERROR = "error"


class MessageDirection(str, Enum):
    REQUEST = "request"
    RESPONSE = "response"


class EventType(str, Enum):
    AGENT_STATUS = "agent.status"
    AGENT_THOUGHT = "agent.thought"
    AGENT_MESSAGE = "agent.message"
    REPORT_CHUNK = "report.chunk"
    REPORT_COMPLETE = "report.complete"
    CHAT_RESPONSE = "chat.response"


# -- Payload models --

class AgentStatusPayload(BaseModel):
    status: AgentStatus
    message: str = ""


class AgentThoughtPayload(BaseModel):
    text: str


class AgentMessagePayload(BaseModel):
    from_agent: str = Field(serialization_alias="from")
    to_agent: str = Field(serialization_alias="to")
    title: str
    description: str = ""
    direction: MessageDirection


class ReportChunkPayload(BaseModel):
    content: str
    section: str
    final: bool = False


class ReportCompletePayload(BaseModel):
    markdown: str
    charts: list[dict] = []


class ChatResponsePayload(BaseModel):
    text: str
    final: bool = False


# -- Envelope --

class SSEEvent(BaseModel):
    event_id: str = Field(default_factory=lambda: str(uuid4()))
    timestamp: float = Field(default_factory=time)
    agent_id: str
    event_type: EventType
    payload: dict

    @classmethod
    def agent_status(cls, agent_id: str, status: AgentStatus, message: str = "") -> "SSEEvent":
        p = AgentStatusPayload(status=status, message=message)
        return cls(agent_id=agent_id, event_type=EventType.AGENT_STATUS, payload=p.model_dump())

    @classmethod
    def agent_thought(cls, agent_id: str, text: str) -> "SSEEvent":
        p = AgentThoughtPayload(text=text)
        return cls(agent_id=agent_id, event_type=EventType.AGENT_THOUGHT, payload=p.model_dump())

    @classmethod
    def agent_message(
        cls,
        agent_id: str,
        from_agent: str,
        to_agent: str,
        title: str,
        description: str = "",
        direction: MessageDirection = MessageDirection.REQUEST,
    ) -> "SSEEvent":
        p = AgentMessagePayload(
            from_agent=from_agent,
            to_agent=to_agent,
            title=title,
            description=description,
            direction=direction,
        )
        return cls(agent_id=agent_id, event_type=EventType.AGENT_MESSAGE, payload=p.model_dump(by_alias=True))

    @classmethod
    def report_chunk(cls, content: str, section: str, final: bool = False) -> "SSEEvent":
        p = ReportChunkPayload(content=content, section=section, final=final)
        return cls(agent_id="orchestrator", event_type=EventType.REPORT_CHUNK, payload=p.model_dump())

    @classmethod
    def report_complete(cls, markdown: str, charts: list[dict] | None = None) -> "SSEEvent":
        p = ReportCompletePayload(markdown=markdown, charts=charts or [])
        return cls(agent_id="orchestrator", event_type=EventType.REPORT_COMPLETE, payload=p.model_dump())

    @classmethod
    def chat_response(cls, text: str, final: bool = False) -> "SSEEvent":
        p = ChatResponsePayload(text=text, final=final)
        return cls(agent_id="orchestrator", event_type=EventType.CHAT_RESPONSE, payload=p.model_dump())

File: frontend/src/schemas/events.ts
import { z } from "zod";

/* ── Agent identifiers ───────────────────────────── */
export const AgentId = z.enum([
  "orchestrator",
  "portfolio",
  "news",
  "modeling",
  "alternatives",
]);
export type AgentId = z.infer<typeof AgentId>;

export const AgentStatus = z.enum(["idle", "working", "done", "error"]);
export type AgentStatus = z.infer<typeof AgentStatus>;

export const MessageDirection = z.enum(["request", "response"]);
export type MessageDirection = z.infer<typeof MessageDirection>;

/* ── Shared data schemas ─────────────────────────── */

export const HoldingSchema = z.object({
  ticker: z.string(),
  weight: z.number(),
  sector: z.string(),
});
export type Holding = z.infer<typeof HoldingSchema>;

export const PortfolioData = z.object({
  sector_allocation: z.record(z.string(), z.number()),
  top_holdings: z.array(HoldingSchema),
  herfindahl_index: z.number(),
  portfolio_beta: z.number(),
});
export type PortfolioData = z.infer<typeof PortfolioData>;

export const HeadlineSchema = z.object({
  title: z.string(),
  sentiment: z.number(),
  ticker: z.string(),
});
export type Headline = z.infer<typeof HeadlineSchema>;

export const NewsData = z.object({
  headlines: z.array(HeadlineSchema),
  aggregate_sentiment: z.record(z.string(), z.number()),
  overall_sentiment: z.number(),
});
export type NewsData = z.infer<typeof NewsData>;

export const ChartOutput = z.object({
  chart_id: z.string(),
  chart_type: z.string(),
  title: z.string(),
  image_base64: z.string(),
  summary: z.string(),
});
export type ChartOutput = z.infer<typeof ChartOutput>;

export const ModelingData = z.object({
  sharpe_ratio: z.number(),
  volatility: z.number(),
  trend_slope: z.number(),
  chart_base64: z.string().nullable().optional(),
  charts: z.array(ChartOutput).optional(),
  metrics: z.record(z.string(), z.number()).optional(),
});
export type ModelingData = z.infer<typeof ModelingData>;

export const AlternativesData = z.object({
  crypto_prices: z.record(z.string(), z.number()),
  cross_correlations: z.record(z.string(), z.number()),
});
export type AlternativesData = z.infer<typeof AlternativesData>;

/* ── Wire format (matches backend SSEEvent envelope) ── */

export const EventType = z.enum([
  "agent.status",
  "agent.thought",
  "agent.message",
  "report.chunk",
  "report.complete",
  "chat.response",
]);
export type EventType = z.infer<typeof EventType>;

export const AgentStatusPayload = z.object({
  status: AgentStatus,
  message: z.string().optional().default(""),
});

export const AgentThoughtPayload = z.object({
  text: z.string(),
});

export const AgentMessagePayload = z.object({
  from: AgentId,
  to: AgentId,
  title: z.string(),
  description: z.string().optional().default(""),
  direction: MessageDirection,
});

export const ReportChunkPayload = z.object({
  content: z.string(),
  section: z.string(),
  final: z.boolean().optional().default(false),
});

export const ReportCompletePayload = z.object({
  markdown: z.string(),
  charts: z.array(ChartOutput).optional().default([]),
});

export const ChatResponsePayload = z.object({
  text: z.string(),
  final: z.boolean().optional().default(false),
});

/** The envelope the backend actually sends over SSE */
export const SSEWireEvent = z.object({
  event_id: z.string(),
  timestamp: z.number(),
  agent_id: z.string(),
  event_type: EventType,
  payload: z.record(z.string(), z.unknown()),
});
export type SSEWireEvent = z.infer<typeof SSEWireEvent>;

/* ── Internal flat types (used by reducer & mock SSE) ── */

export type StatusEvent = {
  agent_id: AgentId;
  type: "status";
  status: AgentStatus;
};

export type ThoughtEvent = {
  agent_id: AgentId;
  type: "thought";
  text: string;
  timestamp?: string;
};

export type ReportSectionEvent = {
  agent_id: AgentId;
  type: "report_section";
  section: "portfolio" | "news" | "modeling" | "alternatives" | "executive_summary";
  data: PortfolioData | NewsData | ModelingData | AlternativesData | { summary: string };
};

export type ChatTokenEvent = {
  agent_id: AgentId;
  type: "chat_response";
  token: string;
  done: boolean;
};

export type SwarmHealthEvent = {
  agent_id: "orchestrator";
  type: "swarm_health";
  active_agents: number;
  signals_per_min: number;
  processing_power: number;
};

export type AgentMessageEvent = {
  agent_id: AgentId;
  type: "agent_message";
  from: AgentId;
  to: AgentId;
  title: string;
  description: string;
  direction: MessageDirection;
};

export type ReportCompleteEvent = {
  agent_id: AgentId;
  type: "report_complete";
  markdown: string;
  charts: ChartOutput[];
};

export type SSEEvent =
  | StatusEvent
  | ThoughtEvent
  | ReportSectionEvent
  | ChatTokenEvent
  | SwarmHealthEvent
  | AgentMessageEvent
  | ReportCompleteEvent;

/* ── Agent metadata (for UI rendering) ──────────── */
export interface AgentMeta {
  id: AgentId;
  name: string;
  icon: string;
  color: string;
}

export const AGENTS: AgentMeta[] = [
  { id: "orchestrator", name: "Orchestrator", icon: "brain", color: "#2563eb" },
  { id: "portfolio", name: "Portfolio Alpha", icon: "pie-chart", color: "#8b5cf6" },
  { id: "news", name: "Sentiment Engine", icon: "newspaper", color: "#f59e0b" },
  { id: "modeling", name: "Quant Modeler", icon: "trending-up", color: "#10b981" },
  { id: "alternatives", name: "Alt Assets", icon: "bitcoin", color: "#ef4444" },
];
